consensus/src/terminator.rs

```diff
@@ -0,0 +1,141 @@
+use std::collections::HashMap;
+
+use futures::channel::oneshot::{channel, Receiver, Sender};
+use log::debug;
+
+type TerminatorConnection = (Sender<()>, Receiver<()>);
+type ShutdownConnection = (Receiver<()>, Option<TerminatorConnection>);
+
+/// Struct that holds connections to offspring and parent components/tasks
+/// and enables a clean/synchronized shutdown
+pub struct Terminator {
+    component_name: &'static str,
+    parent_exit: Receiver<()>,
+    parent_connection: Option<TerminatorConnection>,
+    offspring_exits: HashMap<&'static str, Sender<()>>,
+    offspring_connections: HashMap<&'static str, TerminatorConnection>,
+}
+
+impl Terminator {
+    pub fn new(
+        parent_shutdown_connection: ShutdownConnection,
+        component_name: &'static str,
+    ) -> Self {
+        Self {
+            component_name,
+            parent_exit: parent_shutdown_connection.0,
+            parent_connection: parent_shutdown_connection.1,
+            offspring_exits: HashMap::<_, _>::new(),
+            offspring_connections: HashMap::<_, _>::new(),
+        }
+    }
+
+    /// Creates a terminator for the root component
+    pub fn create_root_terminator(exit: Receiver<()>, name: &'static str) -> Self {
```

Comment:
`create_root` should be ok, since you usually call this as `Terminator::create_root` anyway.